### ISSUE TITLE
Makefile: install prefix violates conventions

### ISSUE THREAD

**Issue by bruce-hill** [neovim/neovim#9615]
In short, the problem is:
```
make
sudo make CMAKE_EXTRA_FLAGS=-DCMAKE_INSTALL_PREFIX=/tmp/neovim-test install
```
The above code will install neovim to `/usr/local`, **not** `/tmp/neovim-test`, and no warnings or error messages will appear.

Normally, either `make && sudo make prefix=/foo install` or `./configure --prefix=/foo && make && sudo make install` would be used (vim supports both), but neovim doesn't follow either convention. Instead, the makefile requires the cryptic (but documented and [googleable](https://github.com/neovim/neovim/issues/2111)) `CMAKE_EXTRA_FLAGS=-DCMAKE_INSTALL_PREFIX=/foo` flag. Unfortunately, the install prefix flag takes effect _during the build process_, rather than during the install process, so it is necessary to completely clean (with `make distclean`, just `make clean` doesn't work) and rebuild neovim in order to change the install directory (which takes several minutes). The build process will also silently install to the original location without warning that the `CMAKE_INSTALL_PREFIX` flag is conflicting. This can be very confusing and it can unintentionally overwrite system files and waste a lot of time.

Ideally, the makefile should support `make && sudo make prefix=/foo install`, like [GNU makefile conventions recommend](https://www.gnu.org/prep/standards/html_node/Directory-Variables.html). But if not, the `CMAKE_INSTALL_PREFIX` flag should at least take effect at install time, not build time.

**Comment by justinmk** [comment]
It's a CMake project. The Makefile is a convenience.

> Normally, either `make && sudo make prefix=/foo install` or `./configure --prefix=/foo && make && sudo make install` 

The latter case is analogous to how the Nvim build works: it requires a configure step.

> The build process will also silently install to the original location without warning that the `CMAKE_INSTALL_PREFIX` flag is conflicting

Yes, let's add a warning (or error). Can you send a patch?

> Ideally, the makefile should support `make && sudo make prefix=/foo install`, like [GNU makefile conventions recommend](https://www.gnu.org/prep/standards/html_node/Directory-Variables.html). But if not, the `CMAKE_INSTALL_PREFIX` flag should at least take effect at install time, not build time.

How?

**Comment by mqudsi** [comment]
For CMake projects with a `make` shortcut/shim, I've taken to adding a `./configure` script that detects the generator (Ninja vs Unix Makefiles) then emits a Makefile that shells out to CMake, with the variables substituted. I don't know if @justinmk would welcome a PR that did that.

**Comment by justinmk** [comment]
@mqudsi 	Thanks, but that seems like a regression because we already have a `Makefile` so users don't need to do a `./configure` step.  Plus we would get tickets if the `./configure` script doesn't behave exactly like a real one ...

AFAICT the most we can feasibly do is error if the current (cached) CMake configuration has a prefix that doesn't match some other attempted prefix.

**Comment by bruce-hill** [comment]
Adding basic support for a `prefix=` flag is pretty simple, it just takes this code below where CMAKE_EXTRA_FLAGS is defined (and maybe updating the README to recommend using it instead):
```
ifneq (,$(prefix))
	CMAKE_EXTRA_FLAGS += -DCMAKE_INSTALL_PREFIX=$(prefix)
endif
```
I don't know the cmake build process well enough to know how to only rebuild the parts related to the install prefix, but this Makefile rule can check if the compiled prefix matches the user-specified prefix (if any) and provide a warning and an opportunity to abort:
```
checkprefix:
	@install_prefix=`echo "$(CMAKE_EXTRA_FLAGS)" | sed -n 's/.*-DCMAKE_INSTALL_PREFIX=\(\([^\\ ]\|\\.\)*\)/\1/p'`; \
	if [ "$$install_prefix" -a -e build/cmake_install.cmake ]; then \
		baked_prefix=`sed -n 's/[ ]*set(CMAKE_INSTALL_PREFIX "\(.*\)")/\1/p' <build/cmake_install.cmake 2>/dev/null`; \
		if [ "$$install_prefix" != "$$baked_prefix" ]; then \
			echo -e "\nWarning! This code was compiled with the install prefix: $$baked_prefix"; \
			echo "But now you've passed in the install prefix: $$install_prefix"; \
			echo "This new value will be ignored."; \
			if [ "`read -P 'Do you want to proceed anyway? [y/N] '`" != "y" ]; then \
				echo "To change the install prefix, run: 'make distclean && make <your prefix flags>'"; \
				exit 1; \
			fi; \
		fi; \
	fi;
```
I made the rule `.PHONY` and added it to the `nvim` dependencies. It could probably stand to be a little more robust, but it detects the basic error case well, and allows the user to choose to proceed anyway. Here's the full Makefile patch: 
[checkprefix_patch.txt](https://github.com/neovim/neovim/files/2871414/checkprefix_patch.txt)

**Comment by justinmk** [comment]
@bruce-hill 	thanks! can you see if #9617 does what you want?

> Adding basic support for a `prefix=` flag is pretty simple, it just takes this code below where CMAKE_EXTRA_FLAGS is defined (and maybe updating the README to recommend using it instead)

I guess we can add it if it's really commonly expected, but I've never seen it before. Where is it documented?

I intentionally didn't add support for `prefix=` because:

- it wouldn't work exactly the way you expected it
- it adds yet another way to set the prefix (besides direct `CMAKE_INSTALL_PREFIX` and indirectly in `CMAKE_INSTALL_PREFIX`)
- it expands the illusion that this is not a CMake project
- it will make the cached-value comparison logic slightly more complicated

**Comment by bruce-hill** [comment]
> can you see if #9617 does what you want?

I can't test it right now, as I'm experiencing some unrelated computer issues, but it looks like it would do the right thing. It might be better to use `sed -n 's/[ ]*set(CMAKE_INSTALL_PREFIX "\(.*\)")/\1/p' <build/cmake_install.cmake 2>/dev/null` instead of `shell $(CMAKE_PRG) -L -N build | 2>/dev/null grep CMAKE_INSTALL_PREFIX  | cut -d '=' -f2`, because fully parsing all the cmake options is kinda slow. The `cmake -L` version is maybe a little more robust, so both are reasonable options though.

> I guess we can add it if it's really commonly expected, but I've never seen it before. Where is it documented? Is this an _autotools_ convention or a _Makefile_ convention?

It's a Makefile convention, [documented here](https://www.gnu.org/prep/standards/html_node/Directory-Variables.html#Directory-Variables). Autoconf also [sets a default or user-specified prefix](https://www.gnu.org/software/autoconf/manual/autoconf.html#Default-Prefix), which can be overridden by passing a new value to `make`. Most make-based projects that I come across support either `prefix=` or `PREFIX=` for setting the install directory. It's not mandatory, but I find it very helpful because I install programs to `/usr/local/stow/<program>-1.2.3` and manage them with `stow`, so having a standardized way to do that makes the process simpler.

> I intentionally didn't add support for `prefix=` because: [...] it adds yet another way to set the prefix [and] it will make the cached-value comparison logic slightly more complicated

I think it's pretty simple to just have `prefix=foo` as an alias for adding `-DCMAKE_INSTALL_PREFIX=foo` to `CMAKE_EXTRA_FLAGS` before `CMAKE_EXTRA_FLAGS` is used anywhere (like in the first part of my patch), and it wouldn't complicate the cached value-checking logic, but...

>it wouldn't work exactly the way you expected it [and] it expands the illusion that this is not a CMake project

those are solid reasons and the fact that the README is very explicit about how to use the CMAKE flag helps a lot. I think it comes down to a judgement call between those reasons vs. having a makefile that follows conventions more closely.

**Comment by mqudsi** [comment]
I looked into the same thing when I wrote [the Makefile CMake shim for fish](https://github.com/fish-shell/fish-shell/blob/master/BSDmakefile) and found the same GNU document about the `prefix` convention. But pretty much every other platform uses `PREFIX` instead, so for the BSD-specific `BSDmakefile` that's what we used.

I coincidentally updated [my GMake proxy project](https://github.com/neosmart/gmake-proxy/commit/5580bf36234c3601ddc1e3465e2324ffe9a081eb) (the same one contributed to this project) just yesterday before this conversation played out to add translation from the uppercased to the lowercased version of `prefix` for the same reasons.
